components/calendar/src/cal/chinese.rs

```diff
@@ -575,47 +574,35 @@ impl<R: Rules> DateFieldsResolver for LunarChinese<R> {
-                }
-            }
-            ComputedOrdinalMonth::NotFound => Err(MonthCodeError::NotInCalendar),
+        // 14 is a sentinel value, greater than all other months, for the purpose of computation only;
+        // it is impossible to actually have 14 months in a year.
+        let leap_month = year.leap_month().unwrap_or(14);
+
+        let (unadjusted @ 1..13, leap) = month_code.to_tuple() else {
+            return Err(MonthCodeError::NotInCalendar);
+        };
+
+        let adjusted = unadjusted + (unadjusted >= leap_month) as u8;
+
+        if leap && unadjusted + 1 == leap_month {
+            return Ok(leap_month);
         }
+
+        if leap && options.overflow != Some(Overflow::Constrain) {
+            return Err(MonthCodeError::NotInYear);
+        }
+
+        Ok(adjusted)
     }
 
-    // The calendar-specific month code represented by `month`;
-    // since the Chinese calendar has leap months, an "L" is appended to the month code for
-    // leap months. For example, in a year where an intercalary month is added after the second
-    // month, the month codes for ordinal months 1, 2, 3, 4, 5 would be "M01", "M02", "M02L", "M03", "M04".
     fn month_code_from_ordinal(&self, year: &Self::YearInfo, ordinal_month: u8) -> ValidMonthCode {
-        // 1 indexed leap month name. This is also the ordinal for the leap month
-        // in the year (e.g. in `M01, M01L, M02, ..`, the leap month is for month 1, and it is also
-        // ordinally `month 2`, zero-indexed)
-        // 14 is a sentinel value
+        // 14 is a sentinel value, greater than all other months, for the purpose of computation only;
+        // it is impossible to actually have 14 months in a year.
         let leap_month = year.leap_month().unwrap_or(14);
-        if leap_month == ordinal_month {
-            // Month cannot be 1 because a year cannot have a leap month before the first actual month,
-            // and the maximum num of months ina leap year is 13.
-            debug_assert!((2..=13).contains(&ordinal_month));
-            ValidMonthCode::new_unchecked(ordinal_month - 1, true)
-        } else {
-            let mut adjusted_ordinal = ordinal_month;
-            if ordinal_month > leap_month {
-                // Before adjusting for leap month, if ordinal > leap_month,
-                // the month cannot be 1 because this implies the leap month is < 1, which is impossible;
-                // cannot be 2 because that implies the leap month is = 1, which is impossible,
-                // and cannot be more than 13 because max number of months in a year is 13.
-                debug_assert!((2..=13).contains(&ordinal_month));
-                adjusted_ordinal -= 1;
-            }
-            debug_assert!((1..=12).contains(&adjusted_ordinal));
-            ValidMonthCode::new_unchecked(adjusted_ordinal, false)
-        }
+        ValidMonthCode::new_unchecked(
+            ordinal_month - (ordinal_month >= leap_month) as u8,
```

Comment:
issue: document please

the previous code is clearer in its behavior even though it's not as compact, IMO.